### ISSUE TITLE
Add MQTTS

### ISSUE THREAD

**Issue by heini66** [Aircoookie/WLED#1346]
**Is your feature request related to a problem? Please describe.**
MQTT isn't secure.

**Describe the solution you'd like**
Add MQTTS and the possibility to apply ssl credentials eg ssl fingerprint

**Describe alternatives you've considered**
There is no alternative to encryption..

**Comment by huggy-d1** [comment]
This is an ESP32-only request - at the moment.  Perhaps eventually it will be a raspberry pi request, or some other microcontroller based system with enough processing power and memory to handle certificates AND certificate management.  Since certs have an end date, they need to get updated in a secure way as well.  All these things drive up the complexity and hardware requirements.

ESP8266 is not capable of handling https/mqtts, and I think most WLED systems are running on ESP8266's, so this request while important in certain applications is important, there are other ways to enhance security.

**Comment by Aircoookie** [comment]
@huggy-d1 is correct. I've attempted making a TLS connection on an ESP8266 and it only works if literally nothing else runs on the system as there is no hardware accelaration and way too little free RAM.

On an ESP32 supporting secure MQTT is possible though  - cert management remains an issue though. Usually arduino sketches just use a hardcoded fingerprint for the specific server they want to contact, which frankly sucks because it expires after a year or so and only is valid for that specific server. The alternative is not verifying the server signature at all, which gives you encryption, however is of limited use since you can't verify the authenticity of the server you are connecting to.

I'm very keen if someone has good ideas on this point :)

**Comment by heini66** [comment]
mqtts as client on an esp8266 is possible .
have a look @ https://github.com/fablab-ka/co2-ampel/blob/master/Software/co2ampel/data/config.json
i would like to have the option to add mqtts credentials in wled to connect to my mqtts broker

**Comment by Aircoookie** [comment]
Of course it is possible, just not with the overhead of hosting a full embedded webserver, filesystem and driving potentially more than 1k individually addressable LEDs. And if it barely works, it would drastically reduce stability.

Security should of course be a given in 2020, but its only possible when there are enough resources. Given that no sensitive data is transmitted, stability is more important for this project at this point.

I would however accept a PR that adds it optionally with a compile time define :)

**Comment by huggy-d1** [comment]
> mqtts as client on an esp8266 is possible .
> have a look @ https://github.com/fablab-ka/co2-ampel/blob/master/Software/co2ampel/data/config.json
> i would like to have the option to add mqtts credentials in wled to connect to my mqtts broker
I just checked your link.  They cheated.  Hard-coded the SSL fingerprint into the firmware.  It is not at all what you think it is, and in a short period of time, your MQTTS server will quite accepting data from an esp8266 using this software.
That said, if your desire to use WLED is seasonal, that solution should work fine.

**Comment by RoganDawes** [comment]
I think there are two possible options that may still be feasible.
1) TLS-PSK (a la mosquitto) - not entirely sure whether there are any libraries that support this on any modern platform, unfortunately.
2) saving the fingerprint of the server's public key (or even the entire public key), not the entire certificate. If you keep the same private and public keypair across certificate renewals, even though the validity period changes, and therefore the signature changes, and thus the fingerprint of the certificate as a whole, you are not giving up a whole lot of security in the process.

**Comment by huggy-d1** [comment]
Creating a self-signed certificate manually so you can use the same key fingerprint for extended periods of time?

**Comment by lhurt** [comment]
Although this request is quite old Tasmota was able to solve this issue and [support MQTT TLS](https://tasmota.github.io/docs/TLS/) since last year using bear ssl library. Any chance to add it to WLED as well?

**Comment by lhurt** [comment]
Is anyone watching this issue?

**Comment by Aircoookie** [comment]
Yes. It's possible on ESP32, but not currently a priority. PR is welcome :)

**Comment by lhurt** [comment]
> Yes. It's possible on ESP32, but not currently a priority. PR is welcome :)

Thank you for explaining. I'd love to contribute, unfortunately it exceeds my capabilities.
BTW: Tasmota offers [MQTTS also for ESP8266](https://tasmota.github.io/docs/TLS/#compiling-tls-for-esp8266).

Thanks btw for your great piece of software.

**Comment by max-koehler** [comment]
Generally speaking, ESP8266 and MQTT TLS can work like a charm. 

The problem is usually establishing the connection, where often out memory issues occur and space issues due to the root certificates that need to be shipped with the firmware. 

I can see that `BearSSL` is already part of the build, and maybe we can scope the feature so that it's easier to start a PR that goes in the right direction.
- Do we have a trust anchors / root certificates that are shipped with `WLED` and loaded into the client already? 
- If not, could we focus on `letsencrypt` only, to safe space and memory?

**Comment by blazoncek** [comment]
> Generally speaking, ESP8266 and MQTT TLS can work like a charm.

That very much depends on the amount of free RAM and available CPU cycles. With WLED installed ESP8266 struggles with both.

**Comment by max-koehler** [comment]
> > Generally speaking, ESP8266 and MQTT TLS can work like a charm.
> 
> That very much depends on the amount of free RAM and available CPU cycles. With WLED installed ESP8266 struggles with both.

In your opinion, it isn't even worth looking into it?

**Comment by blazoncek** [comment]
If your life depends on MQTTS then it may be time to switch to ESP32.
With ESP32 things are more flexible although we are already pushing the limits there too.

**Comment by MateuszPrasal** [comment]
I've prepared my own build for ESP32 with some changes 

1) In platformio.ini - I've changed `marvinroger/AsyncMqttClient @ 0.9.0` to `https://github.com/fAuernigg/async-mqtt-client.git#master` in `env` section (common settings)
2) For `esp32_all_variants` - I've changed `esp32async/AsyncTCP @ 3.4.7` to `https://github.com/tve/AsyncTCP.git#mbed-tls-try2`
3) In  env `esp32dev` I've added `-D ASYNC_TCP_SSL_ENABLED=1` to build flags
4) In `mqtt.cpp` I've added below part of code
```
#ifdef ASYNC_TCP_SSL_ENABLED
  mqtt->setSecure(true);
 #endif
```
Now, I'm testing with HiveMQ MQTT Broker with TLS - works like a charm. 

I didn't push PR because the changes I've made are not so correct - but I need it for some my personal project. Anyway, if someone wants - it works :) 

For sure - I have to add server fingerprint in future, but later - now I'm just testing this solution

**Comment by DedeHai** [comment]
@MateuszPrasal how much RAM and how much flash does that solution use?

**Comment by MateuszPrasal** [comment]
@DedeHai I have to check - I'll let you know

**Comment by MateuszPrasal** [comment]
@DedeHai I've built using `esp32dev` environment. It looks nice, when I used ESP-WROOM-32 with DevKit - I didn't notice any issues

<img width="1092" height="108" alt="Image" src="https://github.com/user-attachments/assets/49ec9985-5e9d-4be6-8b18-ffa48d63d5a6" />

**Comment by DedeHai** [comment]
@MateuszPrasal you need to also post the usage before all your changes.

edit:
also you need to check free-heap before your changes and after. what also is important is heap use during SSL activity, which is a bit harder to determine, it can be done by outputting free heap to Serial in main loop like every 100ms and then monitor that value. Currently a normal web-request uses roughtly 12k of heap.

edit2:
if you are building this on current nightly, your modifications use almost 100k of flash.

**Comment by MateuszPrasal** [comment]
@DedeHai Before changes 

<img width="1076" height="92" alt="Image" src="https://github.com/user-attachments/assets/87f8aa00-5a2e-4154-aa6a-a0f9e24f4b6f" />

Ok, so I need to do some more debugs. I've build `main` branch.

**Comment by DedeHai** [comment]
yea, those 100k of flash are a problem and it is what I have seen in other builds using SSL. I would assume heap use will also increase quite a bit.

**Comment by MateuszPrasal** [comment]
I'll do some more tests :) For my using I can disable most of things like OTA, so I can save lot of memory

**Comment by DedeHai** [comment]
If you get it to work properly, make a branch in your fork and link it here for others to use it.

**Comment by MateuszPrasal** [comment]
Ok, no problem :)

**Comment by MateuszPrasal** [comment]
https://github.com/MateuszPrasal/WLED/tree/mqtts - this is the branch with MQTTS from current `main` branch -  feel free to test :) 

Now, I'm currently do tests on `v0.15.1` tag with 140 LEDs and MQTTS enabled and what I see is that `Free heap` is about 72kB

<img width="458" height="49" alt="Image" src="https://github.com/user-attachments/assets/fb64e9cd-e3e5-413f-aa7b-1c401ce4213c" />

I have also disabled lots of things like websockets, alexa and so on. I'll also test on current main branch so I'll let you know about tests

**Comment by MateuszPrasal** [comment]
I'm also testing on current `main` branch - although there is much less free heap (~25kB) - it's more stable than on v0.15.1. MQTT works much better.
